netket/graph/symmetry.py

```diff
@@ -80,6 +81,62 @@ def remove_duplicates(self, *, return_inverse=False):
         else:
             return group
 
+    def inverse(self):
+        """
+        Returns reordered SymmGroup where the each element is the inverse of
+        the original symmetry element. If :code:`g = self[element]` and :code:`h = self.inverse()[element]`,
+        then :code:`gh = product(g, h)` will act as the identity on the sites of the graph, i.e., :code:`np.all(gh(sites) == sites)`.
+
+        """
+
+        automorphisms = self.to_array()
+        n_symm = len(automorphisms)
+        inverse = np.zeros([n_symm], dtype=int)
+        automorphisms = np.array(automorphisms)
+        for i, perm1 in enumerate(automorphisms):
+            for j, perm2 in enumerate(automorphisms):
+                perm_sq = perm1[perm2]
+                if np.all(perm_sq == np.arange(len(perm_sq))):
+                    inverse[i] = j
+
+        return SymmGroup([self.elems[i] for i in inverse], self.graph)
+
+    def group_algebra(self):
+        """
+        Computes an array of relative displacements between orientations in SymmGroup.
+        So for two symmetry elements, g = self[element] and h = self[element2],
+        self[group_algebra[element1,element2]] = g^{-1} h
+
+        This array is flattened before returning
+
+        """
+
+        automorphisms = self.to_array()
+        n_symm = len(automorphisms)
+        inverse = self.inverse().to_array()
+        group_algebra = np.zeros([n_symm, n_symm], dtype=int)
+
+        inv_t = inverse.transpose()
+        auto_t = automorphisms.transpose()
+        inv_auto = auto_t[inv_t].reshape(-1, n_symm * n_symm).transpose()
+
+        hash_auto = {
+            hash(element.tobytes()): index
+            for index, element in enumerate(automorphisms)
+        }
+
+        inds = [
+            (index, hash_auto[hash(element.tobytes())])
+            for index, element in enumerate(inv_auto)
+            if hash(element.tobytes()) in hash_auto
+        ]
+
+        inds = np.asarray(inds)
+
+        group_algebra[inds[:, 0] // n_symm, inds[:, 0] % n_symm] = inds[:, 1]
+
+        return HashableArray(group_algebra.ravel())
```

Comment:
I think this should return an array. The only place I'd use `HashableArray` is as an attribute of a frozen dataclass (which includes our models). Otherwise it's not worth the hassle.

Reply:
Okay I'll fix this